### PR TITLE
Fix Go to Definition for recursive types

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
This fixes https://github.com/Microsoft/visualfsharp/issues/3304

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs [hunk 1]
```diff
@@ -98,16 +98,21 @@ type internal GoToDefinition(checker: FSharpChecker, projectInfoManager: FSharpP
     member __.FindDefinitionOfSymbolAtRange(targetDocument: Document, symbolRange: range, targetSourceText: SourceText) =
         findSymbolHelper (targetDocument, symbolRange, targetSourceText, false)
     
-    /// use the targetSymbol to find the first instance of its presence in the provided source file
+    /// if the symbol is defined in the given file, ruturn it's declaration location, othewise use the targetSymbol to find the first
```
Comment (smoothdeveloper) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: typo ruturn -> return.

Appreciates the comments BTW

Comment (vasily-kirichenko) [reply vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: Thanks. Fixed.

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs [hunk 2]
```diff
@@ -98,16 +98,21 @@ type internal GoToDefinition(checker: FSharpChecker, projectInfoManager: FSharpP
     member __.FindDefinitionOfSymbolAtRange(targetDocument: Document, symbolRange: range, targetSourceText: SourceText) =
         findSymbolHelper (targetDocument, symbolRange, targetSourceText, false)
     
-    /// use the targetSymbol to find the first instance of its presence in the provided source file
+    /// if the symbol is defined in the given file, return it's declaration location, othewise use the targetSymbol to find the first
```
Comment (cartermp) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: Typos:

it's ---> its
othewise ---> otherwise

**Comment by cartermp** [comment]
Navigation via QuickInfo still behaves as expected, as it does not go through the same code paths

**Comment by vasily-kirichenko** [comment]
@cartermp "It works on my machine" (c)

**Comment by cartermp** [comment]
I'll re-test this when I'm in the office. Could indeed be what you're saying.

**Comment by cartermp** [comment]
"Works on my machine" as well 😄 

Cleaned up my machine a bit and with this installed into 15.7.2, it's working and the repro in #3304 works as well.

**Pull request merged**